common/src/attestations/signers.rs

```diff
@@ -3,161 +3,92 @@
-                        )
-                    }) {
-                    Ok(signer) => {
-                        e.insert(signer);
-                        info!(
-                            "Found attestation signer for {{allocation: {}, deployment: {}}}",
-                            allocation.id, allocation.subgraph_deployment.id,
-                        );
-                    }
-                    Err(e) => {
-                        warn!(
-                        "Failed to find the attestation signer for {{allocation: {}, deployment: {}, createdAtEpoch: {}, err: {}}}",
-                        allocation.id, allocation.subgraph_deployment.id, allocation.created_at_epoch, e
-                    )
-                    }
-                }
-            }
-        }
-    }
-
-    async fn update_loop(inner: Arc<AttestationSignersInner>) {
-        let mut watch_receiver = inner.allocation_monitor.subscribe();
+) -> Eventual<HashMap<Address, AttestationSigner>> {
+    // Keep a cache of the most recent 1000 signers around so we don't need to recreate them
+    // every time there is a small change in the allocations
+    let cache: &'static Mutex<LruCache<_, _>> = Box::leak(Box::new(Mutex::new(LruCache::new(
+        NonZeroUsize::new(1000).unwrap(),
+    ))));
+
+    let indexer_mnemonic = Arc::new(indexer_mnemonic);
+
+    // Whenever the indexer's active or recently closed allocations change, make sure
+    // we have attestation signers for all of them
+    indexer_allocations.map(move |allocations| {
+        let indexer_mnemonic = indexer_mnemonic.clone();
+
+        async move {
+            let mut cache = cache.lock().await;
+
+            for (id, allocation) in allocations.iter() {
+                let result = cache.try_get_or_insert(*id, || {
+                    AttestationSigner::new(&indexer_mnemonic, allocation, chain_id, dispute_manager)
+                });
 
-        loop {
-            match watch_receiver.changed().await {
-                Ok(_) => {
-                    Self::update_attestation_signers(inner.clone()).await;
-                }
-                Err(e) => {
-                    error!(
-                        "Error receiving allocation monitor subscription update: {}",
-                        e
+                if let Err(e) = result {
+                    warn!(
+                        "Failed to establish signer for allocation {}, deployment {}, createdAtEpoch {}: {}",
+                        allocation.id, allocation.subgraph_deployment.id,
+                        allocation.created_at_epoch, e
                     );
                 }
             }
-        }
-    }
 
-    pub async fn read(
-        &self,
-    ) -> tokio::sync::RwLockReadGuard<'_, HashMap<Address, AttestationSigner>> {
-        self.inner.attestation_signers.read().await
-    }
+            HashMap::from_iter(cache.iter().map(|(k, v)| (*k, v.clone())))
+        }
+    })
```

Comment:
This is just a weak suggestion. No big deal.
```suggestion
) -> Eventual<HashMap<Address, AttestationSigner>> {
    let attestation_signers_map: &'static Mutex<HashMap<Address, AttestationSigner>> =
        Box::leak(Box::new(Mutex::new(HashMap::new())));

    let indexer_mnemonic = Arc::new(indexer_mnemonic);

    // Whenever the indexer's active or recently closed allocations change, make sure
    // we have attestation signers for all of them
    indexer_allocations.map(move |allocations| {
        let indexer_mnemonic = indexer_mnemonic.clone();

        async move {
            let mut attestation_signers_map = attestation_signers_map.lock().await;

            // Remove signers for allocations that are no longer active or recently closed
            attestation_signers_map.retain(|id, _| allocations.contains_key(id));

            // Create signers for new allocations
            for (id, allocation) in allocations.iter() {
                if !attestation_signers_map.contains_key(id) {
                    let signer = AttestationSigner::new(
                        &indexer_mnemonic,
                        allocation,
                        chain_id,
                        dispute_manager,
                    );

                    if let Err(e) = signer {
                        warn!(
                            "Failed to establish signer for allocation {}, deployment {}, createdAtEpoch {}: {}",
                            allocation.id, allocation.subgraph_deployment.id,
                            allocation.created_at_epoch, e
                        );
                    } else {
                        attestation_signers_map.insert(*id, signer.unwrap());
                    }
                }
            }

            attestation_signers_map.clone()
        }
    })
```

Reply:
@aasseman Great! I added this with a small variable name change. You're right, the LRU cache isn't needed since we only ever care about retaining the attestation signers for allocations that exist across two runs of this update logic.